src/Mvc/Mvc.Analyzers/src/TagHelpersInCodeBlocksAnalyzer.cs

```diff
@@ -0,0 +1,157 @@
+            {
+                startBlockContext.RegisterOperationAction(context =>
+                {
+                    var awaitOperation = (IAwaitOperation)context.Operation;
+
+                    if (awaitOperation.Operation.Kind != OperationKind.Invocation)
+                    {
+                        return;
+                    }
+
+                    var invocationOperation = (IInvocationOperation)awaitOperation.Operation;
+
+                    if (!IsTagHelperRunnerRunAsync(invocationOperation.TargetMethod, symbolCache))
+                    {
+                        return;
+                    }
+
+                    var parent = context.Operation.Parent;
+                    while (parent != null && !IsParentMethod(parent))
+                    {
+                        parent = parent.Parent;
+                    }
+
+                    if (parent == null)
+                    {
+                        return;
+                    }
+
+                    var methodSymbol = (IMethodSymbol)(parent switch
+                    {
+                        ILocalFunctionOperation localFunctionOperation => localFunctionOperation.Symbol,
+                        IAnonymousFunctionOperation anonymousFunctionOperation => anonymousFunctionOperation.Symbol,
+                        IMethodBodyOperation methodBodyOperation => startBlockContext.OwningSymbol,
+                        _ => null,
+                    });
+
+                    if (methodSymbol == null)
+                    {
+                        // Unsupported operation type.
+                        return;
+                    }
+
+                    if (!methodSymbol.IsAsync ||
+                        !symbolCache.TaskType.IsAssignableFrom(methodSymbol.ReturnType))
+                    {
+                        if (parent.Syntax.GetDiagnostics().Any(diagnostic => diagnostic.Id == TagHelperInCodeBlockDiagnostic.Id))
+                        {
+                            // Diagnostic has already been logged on this syntax item, no-op.
+                            return;
+                        }
+
+                        context.ReportDiagnostic(
+                            Diagnostic.Create(
+                                TagHelperInCodeBlockDiagnostic,
+                                parent.Syntax.GetLocation()));
+                    }
+
+                }, OperationKind.Await);
+            });
+        }
+
+        private bool IsTagHelperRunnerRunAsync(IMethodSymbol method, SymbolCache symbolCache)
+        {
+            if (method.IsGenericMethod)
+            {
+                return false;
+            }
+
+            if (method != symbolCache.TagHelperRunnerRunAsyncMethodSymbol)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private bool IsParentMethod(IOperation operation)
+        {
+            if (operation.Kind == OperationKind.LocalFunction)
```

Comment:
Wouldn't the `OperationKind.MethodBody` catch this? LocalFunctions would have bodies too, no?

Reply:
Talked offline, gist, no it's different 😄